### PR TITLE
Fix for Arbitrary File Read in gitserver using diff command

### PR DESCRIPTION
**Pull request by shivasurya**
Resolves: https://github.com/sourcegraph/security-issues/issues/348

Where paths could be added after --. This is not exploitable by regular users, but can be when HTTP endpoints are exposed over HTTP and gitserver is called directly.

## Test plan

- [x] Check out branch and try to run PoC again. (verified)
- [x] added testcase

### CODE REVIEW

File: internal/gitserver/gitdomain/exec.go [hunk 1]
```diff
@@ -103,10 +102,12 @@ func isAllowedDiffArg(arg string) bool {
 		}
 	}
 
-	// make sure that arg is not a local file
-	_, err := os.Stat(arg)
+	if strings.HasPrefix(arg, "/") || strings.HasPrefix(arg, "../") {
```
Comment (evict) [inline internal/gitserver/gitdomain/exec.go:105]: This works, but we can also resolve the path, then we can make sure we are in the right directory at all times. Ideally we'd enforce this on the repo level. If that's not possible then this should work too:

```
// resolve the full path 
absPath, err := filepath.Abs(arg)

if !strings.HasPrefix(absPath, "/data/repos")  {
 // fail here
```

The current working directory should be pointing at the repo, so we could improve the above check a bit.

Comment (eseliger) [reply internal/gitserver/gitdomain/exec.go:105]: Agree with that, also, I think using 
```
cmd/../../../etc/passwd
```
would currently work, which @evict s suggestion also fixes. (can we get a test case for this added please?)

File: internal/gitserver/gitdomain/exec_test.go [hunk 2]
```diff
@@ -53,6 +53,10 @@ func TestIsAllowedDiffGitCmd(t *testing.T) {
 		{args: []string{"diff", "/etc/hosts", "/etc/passwd"}, pass: false},
 		{args: []string{"diff", "/dev/null", "/etc/passwd"}, pass: false},
 		{args: []string{"diff", "/etc/hosts", "/etc/passwd"}, pass: false},
+		{args: []string{"diff", "--", "/etc/hosts", "/etc/passwd"}, pass: false},
+		{args: []string{"diff", "--", "/etc/ees", "/etc/ee"}, pass: true},
```
Comment (eseliger) [inline internal/gitserver/gitdomain/exec_test.go:57]: why should this one be allowed? 🤔

File: internal/gitserver/gitdomain/exec.go (hunk 1 above)
Comment (shivasurya) [reply internal/gitserver/gitdomain/exec.go:105]: @evict @eseliger 

is there a way to get the current repo directory? if not it's hard to verify the `!strings.HasPrefix(absPath, "/data/repos")`. The os.Getwd() returns directory golang script path than current repo.

File: internal/gitserver/gitdomain/exec_test.go (hunk 2 above)
Comment (shivasurya) [reply internal/gitserver/gitdomain/exec_test.go:57]: this testcase is for verifying non existing path in the disk.

**Review by evict — APPROVED** [review]

File: cmd/gitserver/server/patch.go [hunk 3]
```diff
@@ -62,6 +62,7 @@ func (s *Server) createCommitFromPatch(ctx context.Context, req protocol.CreateC
 	var resp protocol.CreateCommitFromPatchResponse
 
 	repo := string(protocol.NormalizeRepo(req.Repo))
+	repoDir := filepath.Join(s.ReposDir, repo)
 	repoGitDir := filepath.Join(s.ReposDir, repo, ".git")
```
Comment (eseliger) [inline cmd/gitserver/server/patch.go:66]: makes their relation a bit clearer imo
```suggestion
	repoGitDir := filepath.Join(repoDir, ".git")
```

**Comment by mrnugget** [comment]
@shivasurya can you please give this a proper PR title?

File: internal/gitserver/gitdomain/exec.go [hunk 4]
```diff
@@ -103,9 +105,7 @@ func isAllowedDiffArg(arg string) bool {
 		}
 	}
 
-	// make sure that arg is not a local file
 	_, err := os.Stat(arg)
-
```
Comment (eseliger) [inline internal/gitserver/gitdomain/exec.go:108]: why not keep this comment? helps me understand why we have it 😬

File: internal/gitserver/gitdomain/exec.go [hunk 5]
```diff
@@ -2,13 +2,15 @@ package gitdomain
...
+	"path/filepath"
 	"strconv"
 	"strings"
 
 	"github.com/grafana/regexp"
-	"k8s.io/utils/strings/slices"
 
 	"github.com/sourcegraph/log"
+
+	"k8s.io/utils/strings/slices"
```
Comment (eseliger) [inline internal/gitserver/gitdomain/exec.go:13]: wonder why this reordering happened 🤔

File: internal/gitserver/gitdomain/exec.go [hunk 6]
```diff
@@ -128,8 +128,34 @@ func isAllowedGitArg(allowedArgs []string, arg string) bool {
...
+	if err != nil {
+		return false
+	} else {
+		// Check if absolute path is a sub path of the repo dir
+		repoRoot, err := filepath.Abs(repoDir)
+		if err == nil {
+			return strings.HasPrefix(filePath, repoRoot)
+		}
+	}
+	return false
```
Comment (eseliger) [inline internal/gitserver/gitdomain/exec.go:154]: this feels easier to read to me, wdyt? the else branch after a return is not necessary, for example :)

```suggestion
	arg = filepath.Clean(arg)
	if !filepath.IsAbs(arg) {
		arg = filepath.Join(repoDir, arg)
	}
	filePath, err := filepath.Abs(arg)
	if err != nil {
		return false
	}
	
		// Check if absolute path is a sub path of the repo dir
		repoRoot, err := filepath.Abs(repoDir)
		if err != nil {
		return false
		}
			return strings.HasPrefix(filePath, repoRoot)
```

**Review by eseliger — APPROVED** [review]
Did you manually go through a few SG features to verify they still work alright? 

I would want to ask if we can defer merging this PR until Monday when people are awake, in case something unforeseen happens :)

**Comment by shivasurya** [comment]
Thanks @eseliger , i'll fix those suggestions and wait till monday for the merge 👍🏻

**Pull request merged**